windows/src/desktop/kmshell/main/initprog.pas

```diff
@@ -698,5 +652,76 @@ procedure PrintKeyboard(KeyboardName: WideString);  // I2329
     end;
 end;
 
+function ProcessBackroundUpdate(FMode: TKMShellMode; FSilent: Boolean) : Boolean;
+var
+  BUpdateSM : TUpdateStateMachine;
+  frmStartInstall: TfrmStartInstall;
+  UserCanceled : Boolean;
+  SkipBUpdate : Boolean;
+begin
+  BUpdateSM := TUpdateStateMachine.Create(False);
+  Result := False;
+    try
+      if (FMode = fmBackgroundUpdateCheck) then
+      begin
+        BUpdateSM.HandleCheck;
+        Result := True;
+        Exit;
+      end
+      else if (FMode = fmBackgroundDownload) then
+      begin
+        BUpdateSM.HandleDownload;
+        Result := True;
+        Exit;
+      end
+      else if (FMode = fmApplyInstallNow) then
+      begin
+        BUpdateSM.HandleInstallNow;
+        Result := True;
+        Exit;
+      end
+      else if (FMode = fmInstallKeyboardPackageAdmin) then
+      begin
+        BUpdateSM.HandleInstallPackages;
+        Result := True;
+        Exit;
+      end
+      else
+      begin
+        // Since Package upgrade and Keyman upgrade share the installing state
+        // the following package related switches are valid in in the installing
+        // state.
+        SkipBUpdate := (BUpdateSM.IsInstallingState and (FMode in [fmInstallTip,
+        fmInstallTipsForPackages, fmRegisterTip, fmUpgradeKeyboards,
+        fmUpgradeMnemonicLayout]));
+        // The following logic around the WaitingRestartState should be
+        // encapsulated in the state machine however as we want separation of
+        // UI elements from the state machine we have bring some of logic here.
+        UserCanceled := False;
+        if BUpdateSM.ReadyToInstall and
+          (not FSilent and (FMode in [fmStart, fmSplash, fmMain, fmAbout, fmHelp,
+          fmShowHelp, fmSettings])) then
```

Comment:
@marc I have added fmShowHelp, and fmSettings as gui's where we could notify the user we can pop keyman update is ready to install. select  `update`, or `close`.

However, I am still not happy that we are missing the all-important case when Keyman starts with Windows which is simply `kmshell.exe -s`   I propose we update this to have a switch `-on-boot`  so `kmshell.exe -s -on-boot` then we could determine this is a case where the UI can be run. Thoughts?

Reply:
> @marc

This isn't me!

Reply:
`kmshell -s` means 'silent' so that means no UI is intended to be shown there... We could change that to `kmshell.exe -boot`, but we'd have to fixup existing installations -- we'd need to make sure that that change was applied during the upgrade (this is quite doable, just more work and testing).

I think that'd be good because it'd be clearer than the current `-s` which could be used any time and is ambiguous.

Reply:
> > @_marc
> 
> This isn't me!

yikes